knowhere/feder/HNSW.h

```diff
@@ -0,0 +1,203 @@
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under the License
+
+#pragma once
+
+#include <memory>
+#include <tuple>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include <nlohmann/json.hpp>
+
+#include "knowhere/feder/IDSet.h"
+
+namespace knowhere::feder::hnsw {
+
+///////////////////////////////////////////////////////////////////////////////
+// index view
+struct NodeInfo {
+    int64_t id_;
+    std::vector<int64_t> links_;
+
+    NodeInfo() = default;
+
+    NodeInfo(const int64_t id, const std::vector<int64_t>& links) = delete;
+
+    NodeInfo(const int64_t id, std::vector<int64_t>&& links) : id_(id), links_(links) {
+    }
+
+    NodeInfo(const NodeInfo& other) {
+        id_ = other.id_;
+        links_ = other.links_;
+    }
+
+    NodeInfo(NodeInfo&& other) {
+        id_ = other.id_;
+        links_ = std::move(other.links_);
+    }
+
+    NodeInfo
+    operator=(const NodeInfo& other) {
+        id_ = other.id_;
+        links_ = other.links_;
+        return *this;
+    }
+
+    NodeInfo
+    operator=(NodeInfo&& other) {
+        id_ = other.id_;
+        links_ = std::move(other.links_);
+        return *this;
+    }
+
+    NLOHMANN_DEFINE_TYPE_INTRUSIVE(NodeInfo, id_, links_)
+};
+
+struct LevelLinkGraph {
+    int64_t level_;
+    std::vector<NodeInfo> nodes_;
+
+    LevelLinkGraph() = default;
+
+    explicit LevelLinkGraph(int64_t level) : level_(level) {
+    }
+
+    LevelLinkGraph(const LevelLinkGraph& other) {
+        level_ = other.level_;
+        nodes_ = other.nodes_;
+    }
+
+    LevelLinkGraph(LevelLinkGraph&& other) {
+        level_ = other.level_;
+        nodes_ = std::move(other.nodes_);
+    }
+
+    LevelLinkGraph
+    operator=(const LevelLinkGraph& other) {
+        level_ = other.level_;
+        nodes_ = other.nodes_;
+        return *this;
+    }
+
+    LevelLinkGraph
+    operator=(LevelLinkGraph&& other) {
+        level_ = other.level_;
+        nodes_ = std::move(other.nodes_);
+        return *this;
+    }
+
+    void
+    AddNodeInfo(int64_t id, std::vector<int64_t>& links) {
+        nodes_.emplace_back(NodeInfo(id, std::move(links)));
+    }
+
+    NLOHMANN_DEFINE_TYPE_INTRUSIVE(LevelLinkGraph, level_, nodes_)
+};
+
+struct HNSWMeta {
+    int64_t ef_construction_;
+    int64_t M_;
+
+    int64_t num_elem_;
+    int64_t num_levels_;
+
+    int64_t enter_point_id_;
+    int64_t num_overview_levels_;
+
+    std::vector<LevelLinkGraph> overview_hier_graph_;
+
+    void
+    AddLevelLinkGraph(int64_t level) {
+        overview_hier_graph_.emplace_back(LevelLinkGraph(level));
+    }
```

Comment:
Do we need the level? From the overview, looks like we only allow use to add `NodeInfo` to the latest level, so why we need use to give us a level id?

Reply:
"level id" is needed by front end